### PR TITLE
Added `Builder` class to build rules

### LINKED ISSUE #7: Add ability to create rules programmatically
This would enable conditional rule creation without relying on string manipulation and can also increase readability of rules written in code.

### PR DESCRIPTION
**Pull request by ricardoboss**
Fixes #7

### CODE REVIEW

File: PasswordRulesSharp.Tests/Builder/BuilderTests.cs [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+
+    [Test]
+    public void BuildComplex()
+    {
+        var rule = new PasswordRulesSharp.Builder.Builder()
+            .MinLength(8)
+            .MaxLength(128)
+            .MaxConsecutive(5)
+            .RequireDigit()
+            .RequireAnyOf(CharacterClass.Lower, CharacterClass.Upper)
```
Comment (chucker) [inline PasswordRulesSharp.Tests/Builder/BuilderTests.cs:30]: neat!

**Comment by chucker** [comment]
I like it. Will have to play with it.

**Comment by ricardoboss** [comment]
I played with the idea of refactoring `Rule` to `StringRule` or something like that, but it would break the existing API.

**Comment by chucker** [comment]
> I played with the idea of refactoring `Rule` to `StringRule` or something like that, but it would break the existing API.

I'm OK with that, but I think I would prefer

* not having a `BaseRule`; call that `Rule`
* `Rule` (currently `BaseRule`)'s constructor being `internal`
* `Rule` having a static method `FromString`

**Comment by ricardoboss** [comment]
When adding breaking changes, do you mind also refactoring all `Rule` classes into their own namespace (`PasswordSharpRules.Rules`)?

**Comment by chucker** [comment]
> When adding breaking changes, do you mind also refactoring all `Rule` classes into their own namespace (`PasswordSharpRules.Rules`)?

Yeah, sounds good.

(`Tokenizer` should probably also be `internal`, and maybe move to a new namespace `PasswordSharpRules.Rules.Parsing`?)

**Comment by ricardoboss** [comment]
I wouldn't want to touch the tokenizer, as it seems OOS for this PR

**Comment by chucker** [comment]
> I wouldn't want to touch the tokenizer, as it seems out of scope for this PR

c9e3732c6af01c2618c9da7cd10c978fca26236a

**Comment by ricardoboss** [comment]
@chucker Done

**Comment by ricardoboss** [comment]
The only thing left that I would change is to make `IncompleteRule` internal too.

**Comment by chucker** [comment]
> The only thing left that I would change is to make `IncompleteRule` internal too.

agreed

File: PasswordRulesSharp/Builder/Builder.cs [hunk 2]
```diff
@@ -0,0 +1,86 @@
+using System.Linq;
+using NodaTime;
+using PasswordRulesSharp.Models;
+using PasswordRulesSharp.Parser;
+
+namespace PasswordRulesSharp.Builder;
+
+public class Builder
```
[the remark this answers is not in the document]
Comment (chucker) [reply PasswordRulesSharp/Builder/Builder.cs:8]: ^

**Review by chucker — CHANGES_REQUESTED** [review]

File: PasswordRulesSharp/Builder/Builder.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (ricardoboss) [reply PasswordRulesSharp/Builder/Builder.cs]: Yes.

**Review by chucker — APPROVED** [review]

**Pull request merged**